# Exploring eBay Car Sales Data

## Introduction

In this project, we will clean the data to analyse used car listings. The data sampled from eBay Germany contains 50,000 data points. The original dataset can be found [here](https://www.kaggle.com/orgesleka/used-cars-database/data).

The data dictionary is as follows:

- **dateCrawled:** The date when the ad was first crawled
- **name:** Name of the car
- **seller:** Whether the seller is private or a dealer
- **offerType:** The type of listing
- **price:** The price advertised price of the car
- **abtest:** Whether the listing is included in an A/B test
- **vehicleType:** The type of vehicle
- **yearOfRegistration:** The year in which the car was first registered
- **gearbox:** The transmission type
- **powerPS:** The power of the car in PS
- **model:** The car model name
- **kilometer:** How many km's the car has driven
- **monthOfRegistration:** The month in which the car was first registered
- **fuelType:** What type of fuel the car uses
- **brand:** The brand of the car
- **notRepairedDamage:** If the car has a damage which is not yet repaired
- **dateCreated:** The date on which the eBay listing was created
- **nrOfPictures:** The number of pictures in the ad
- **postalCode:** The postcode for the location of the vehicle
- **lastSeenOnline:** When the crawler last saw the ad online

## Data Load

For this project, we import the pandas and NumPy libraries. 

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv("autos.csv",encoding = 'Latin-1')

### Column Name Standardisation

We begin by printing an array of the existing column names.

In [3]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

We then copy this array and will make the following changes to the column names:

- yearOfRegistration will become registration_year
- monthOfRegistration will become registration_month
- notRepairedDamage will become unrepaired_damage
- dateCreated will become ad_created
- dateCrawled will become date_crawled

We will also convert all remaining column names to lower case.

In [4]:
'''
Renaming columns
'''

autos.rename(columns={"yearOfRegistration": "registration_year"}, inplace = True)
autos.rename(columns={"monthOfRegistration": "registration_month"}, inplace = True)
autos.rename(columns={"notRepairedDamage": "unrepaired_damage"}, inplace = True)
autos.rename(columns={"dateCreated": "ad_created"}, inplace = True)
autos.rename(columns={"dateCrawled": "date_crawled"}, inplace = True)
autos.rename(columns={"lastSeen": "last_seen"}, inplace = True)

'''
Converting all column headers to lower case
'''
autos.columns = map(str.lower,autos.columns)

autos.columns # Check that this has been completed correctly

Index(['date_crawled', 'name', 'seller', 'offertype', 'price', 'abtest',
       'vehicletype', 'registration_year', 'gearbox', 'powerps', 'model',
       'odometer', 'registration_month', 'fueltype', 'brand',
       'unrepaired_damage', 'ad_created', 'nrofpictures', 'postalcode',
       'last_seen'],
      dtype='object')

The column headers needed to be changed for a number of reasons:

- **Simplicity:** Short column names improve readability and make it easier to perform operations in panda. 

- **Standardisation:** By converting all column names to lowercase, we don't have to remember where the capital letters are located in the string when working with Pandas

## Data Exploration & Observations

As this was successfully loaded and the columns standardised, we will now explore the data.

In [5]:
print(autos.info()) #This will provide information about the dataframe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          50000 non-null object
name                  50000 non-null object
seller                50000 non-null object
offertype             50000 non-null object
price                 50000 non-null object
abtest                50000 non-null object
vehicletype           44905 non-null object
registration_year     50000 non-null int64
gearbox               47320 non-null object
powerps               50000 non-null int64
model                 47242 non-null object
odometer              50000 non-null object
registration_month    50000 non-null int64
fueltype              45518 non-null object
brand                 50000 non-null object
unrepaired_damage     40171 non-null object
ad_created            50000 non-null object
nrofpictures          50000 non-null int64
postalcode            50000 non-null int64
last_seen             50000 non-null object
dtypes: int64(5)

In [6]:
autos.head() #This will provide a sample of the first few rows

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offertype   price   abtest vehicletype  registration_year  \
0  privat   Angebot  $5,000  control         bus               2004   
1  privat   Angebot  $8,500  control   limousine               1997   
2  privat   Angebot  $8,990     test   limousine               2009   
3  privat   Angebot  $4,350  control  kleinwagen               2007   
4  privat   Angebot  $1,350     test       kombi               2003   

     gearbox  powerps   model   odometer  registration_month fueltype  \
0    manuell      158  andere  150,000km                   3      lpg   
1  automatik      286     7er  150,000km                   6   benzin   
2    manuell      102    golf   70,000km                   7   benzin   
3  automatik       71  fortwo   70,000km                   6   benzin   
4    manuell        0   focus  150,000km                   7   benzin   

        brand unrepaired_damage           ad_created  nrofpictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalcode            last_seen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [7]:
autos.describe(include="all") # Obtain description statistics

date_crawled         name  seller offertype  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-30 17:37:35  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicletype  registration_year  gearbox       powerps  model  \
count        44905       50000.000000    47320  50000.000000  47242   
unique           8                NaN        2           NaN    245   
top      limousine                NaN  manuell           NaN   golf   
freq         12859                NaN    36993           NaN   4024   
mean           NaN        2005.073280      NaN    116.355920    NaN   
std            NaN         105.712813      NaN    209.216627    NaN   
min            NaN        1000.000000      NaN      0.000000    NaN   
25%            NaN        1999.000000      NaN     70.000000    NaN   
50%            NaN        2003.000000      NaN    105.000000    NaN   
75%            NaN        2008.000000      NaN    150.000000    NaN   
max            NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fueltype       brand unrepaired_damage  \
count       50000        50000.000000    45518       50000             40171   
unique         13                 NaN        7          40                 2   
top     150,000km                 NaN   benzin  volkswagen              nein   
freq        32424                 NaN    30107       10687             35232   
mean          NaN            5.723360      NaN         NaN               NaN   
std           NaN            3.711984      NaN         NaN               NaN   
min           NaN            0.000000      NaN         NaN               NaN   
25%           NaN            3.000000      NaN         NaN               NaN   
50%           NaN            6.000000      NaN         NaN               NaN   
75%           NaN            9.000000      NaN         NaN               NaN   
max           NaN           12.000000      NaN         NaN               NaN   

                 ad_created  nrofpictures    postalcode            last_seen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

### Observations

Based on the descriptive statistics above, I've made some obserations:


**Columns to change the data type**

- **price**: This needs to be converted into either an integer or a float.

- **odomoter**: This column contains values stored as text, we will need to remove the 'km' after the figure. 

**Columns with 'dirty' data**

- **powerps**: The minimum value is 0 and the maximum is 17,700. This seems like an odd range.

- **registration_year**: This has a max of 9999 and a min of 1000 which is far outside a possible range.

- **registration_month**: The minimum value is 0. Does this mean that the car is unregistered?

**Columns to drop**

- **seller**: This column contains 2 values of which 1 is not 'privat'. This column can be dropped.

- **offer_type**: This column contains 2 values of which 1 is not 'Angebot'. This column can be dropped. 

- **nrofpictures**: This column is completely empty. This column can be dropped.

- **date_crawled, last_seen, ad_created**: The data in these columns for this project are not required as they don't offer any useful datetime info. We will remove these columns

Let's look at some of these columns in greater detail.

In [8]:
'''
Registration years
'''

print(autos["registration_year"].value_counts().sort_index(ascending=False).head(10))
print("\n")
print(autos["registration_year"].value_counts().sort_index(ascending=True).head(10))

9999    4
9996    1
9000    2
8888    1
6200    1
5911    1
5000    4
4800    1
4500    1
4100    1
Name: registration_year, dtype: int64


1000    1
1001    1
1111    1
1500    1
1800    2
1910    9
1927    1
1929    1
1931    1
1934    2
Name: registration_year, dtype: int64


From the above output, we see that a number of registration years are incorrect.  We will need to identify these cars later on. 

In [9]:
'''
Registration month
'''
autos["registration_month"].value_counts().sort_index()

0     5075
1     3282
2     3008
3     5071
4     4102
5     4107
6     4368
7     3949
8     3191
9     3389
10    3651
11    3360
12    3447
Name: registration_month, dtype: int64

From the registration month data, we can see that a number of cars are registered in month 0 which is incorrect. There is a line-ball decision to make, while the month is incorrect, what month the vehicle is registered isn't very important to our analysis. Cars registered in month 0 make up approximately 10% of all listings and the largest group of car registrations. In this instance, we will leave these listings as is.

In [10]:
'''
Number of pictures
'''

autos["nrofpictures"].value_counts()

0    50000
Name: nrofpictures, dtype: int64

This column will need to be discarded as it won't provide any analytical information.

In [11]:
'''
Power Values
'''
print(autos["powerps"].value_counts().sort_index(ascending=False).head(10))
print("\n")
print(autos["powerps"].value_counts().sort_index(ascending=True).head(10))

17700    1
16312    1
16011    1
15016    1
15001    1
14009    1
9011     1
8404     1
7511     1
6512     1
Name: powerps, dtype: int64


0     5500
1        5
2        2
3        3
4        4
5       13
6        3
8        2
9        1
10       3
Name: powerps, dtype: int64


In laymans terms PS is the equivalent of Horsepower. There is quite a large range of power outputs. However we should be concerned with the outliers such as cars with a powerps value of 0 and perhaps those with power outputs greater than say 1100 ps.

## Cleaning the Data

#### Converting Numeric Values Stored as Text

We need to convert the values in the price and odometer columns into numeric values. To do so we will:

- Remove any non-numeric characters
- Convert the whole column to a numeric dtype
- Rename the column so we know what the unit of measurement is

In [12]:
'''
Converting the price column to numeric value
'''

autos["price"] = autos["price"].str.replace('$','').str.replace(",","")
# remove the $ and , values from each figure

autos["price"] = autos["price"].astype(int)
# convert the whole column to an integer

'''
Converting the odometer column to a numeric value
'''

autos["odometer"] = autos["odometer"].str.replace('kms','').str.replace('km','').str.replace(',','')
# remove either 'kms', 'km' or any comma

autos["odometer"] = autos["odometer"].astype(int)
# convert the whole column to an interger

autos.rename(columns={"odometer": "odometer_km"}, inplace = True)
# rename the odometer column to unclude the UoM.

autos.dtypes
# check that the price and odometer columns now contain integers

date_crawled          object
name                  object
seller                object
offertype             object
price                  int64
abtest                object
vehicletype           object
registration_year      int64
gearbox               object
powerps                int64
model                 object
odometer_km            int64
registration_month     int64
fueltype              object
brand                 object
unrepaired_damage     object
ad_created            object
nrofpictures           int64
postalcode             int64
last_seen             object
dtype: object

We can see from the above that the odometer and price columns now contain numeric values.

### Identifying and Eliminating Outliers

As we have converted the odometer and price columns to be integers, we can identify the outliers. We will also identify outliers in the registration column.

In [13]:
'''
Identify unique values
'''

print(autos["odometer_km"].unique().shape)
print(autos["price"].unique().shape)
print(autos["registration_year"].unique().shape)
print(autos["powerps"].unique().shape)


(13,)
(2357,)
(97,)
(448,)


We can see that the odometer column contains 13 unique values which suggests that this comes from a pre-defined range. 

Price on the other hand has 2,357 unique values so the range of prices is quite wide.

Registration year contains 97 unique values which seem ok on the surface but as we know, there are definitely some questionable items there. 

The powerps column contains 436 unique values however as above, there are definitely some questionable items there.

In [14]:
'''
Series descriptions
'''

print(autos["odometer_km"].describe())
print("\n")
print(autos["price"].describe())
print("\n")
print(autos["registration_year"].describe())
print("\n")
print(autos["powerps"].describe())

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64


count    50000.000000
mean       116.355920
std        209.216627
min          0.000000
25%         70.000000
50%        105.000000
75%        150.000000
max      17700.000000
Name: powerps, dtype: float64


The data descriptions show that the odometer values appear to be quite clean while the price, registration year and powerps values are quite extreme. We will now identify the outliers.

#### Odometer Outliers

In [15]:
#Value counts for odometer

autos["odometer_km"].value_counts().sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

The odometer values appear fine so we will not be cleaning them. Let's take a look at the price values.

#### Price Outliers

In [16]:
#Value counts for price

print(autos["price"].value_counts().sort_index(ascending=False).head(5))
print("\n")
print(autos["price"].value_counts().sort_index(ascending=True).head(5))

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
Name: price, dtype: int64


0    1421
1     156
2       3
3       1
5       2
Name: price, dtype: int64


From the above, cars with a price of under 100 and cars with a price tag of more than 200,000 are worth exploring further. Lets take a look.

In [17]:
autos.loc[autos["price"] < 100] # Cars with a price less than $100

date_crawled                                               name  \
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
30     2016-03-14 11:47:31                         Peugeot_206_Unfallfahrzeug   
55     2016-03-07 02:47:54                     Mercedes_E320_AMG_zu_Tauschen!   
64     2016-04-05 07:36:19                 Autotransport__Abschlepp_Schlepper   
71     2016-03-28 19:39:35  Suche_Opel_Astra_F__Corsa_oder_Kadett_E_mit_Re...   
80     2016-03-09 15:57:57  Nissan_Primera_Hatchback_1_6_16v_73_Kw___99Ps_...   
87     2016-03-29 23:37:22                      Bmw_520_e39_zum_ausschlachten   
99     2016-04-05 09:48:54                    Peugeot_207_CC___Cabrio_Bj_2011   
118    2016-03-12 05:03:00  VW_Sharan_V6_204_PS_Karosse_Rohkarosse_mit_Pap...   
146    2016-03-22 23:59:28                                    Ford_Fiesta_rot   
167    2016-04-02 19:43:45  Suche_VW_Multivan_Innenausstattung_Set_oder_TE...   
180    2016-03-19 10:50:25                                       Zu_verkaufen   
226    2016-03-25 23:52:12                      Porsche_911_S_Targa__67er_SWB   
234    2016-03-05 16:56:22                                   Fiat_Punto_Sport   
245    2016-03-30 12:45:44                                     Renault_Twingo   
248    2016-03-08 22:57:53                             VW_Passat_zu_verkaufen   
259    2016-04-03 23:49:58                      guenstiges_Auto_/_auch_defekt   
301    2016-03-08 20:37:59                         Kaufe_alle_Autos_bietet_an   
323    2016-03-15 16:54:59                                W220_Unfallfahrzeug   
327    2016-04-03 13:49:21              BMW_Cabrio_318i_/_BMW_318_tds_Touring   
344    2016-03-22 11:25:18  Verkaufe_hier_mein_mein_schoener_honda_crx_in_...   
347    2016-04-02 11:48:02                                  Volvo_TP21__Sugga   
366    2016-03-21 20:36:25  Ford_Escort_Express_75__LKW_Zulussung___Diesel...   
389    2016-03-08 11:38:52                                    Vw_polo_classic   
418    2016-03-29 14:43:24         Fiat_SCUDO_8_Sitzer_Bus__Diesel_JTD__80_KW   
427    2016-03-12 11:44:47                 TOP_ZUSTAND_GrunePlakette_Euro4!!!   
430    2016-03-18 23:52:40                                  Winterraeder_FORD   
449    2016-04-03 20:59:59                           Tausche_gegen_ein_smart.   
487    2016-03-14 21:41:36  MAG_MICH_KEINER_HABEN_Opel_Omega_B_Caravan_2.0...   
...                    ...                                                ...   
49082  2016-03-13 17:06:58         Passat_35i_b4_in_candy_weiss_original_lack   
49240  2016-03-10 09:54:48   Suche_7_Sitzer_Espace_oder?_Biete_s.Beschreibung   
49254  2016-03-21 10:49:57                           Corrado_G60_mit_2e_Motor   
49301  2016-03-20 20:52:40                           Golf_3_Variant_Kombi_1.8   
49342  2016-03-09 20:47:01  Audi_80_b4_S_Color_und_BMW_318ti_compact_Tausc...   
49350  2016-03-19 23:57:15                                     Golf_II_Diesel   
49377  2016-03-20 00:58:36             Tausche_BMW_318i_Facelift_Model_143_ps   
49404  2016-04-04 03:03:16                                Golf_2_BBS_LEDER_KW   
49408  2016-03-30 20:57:06                                         OPEL_ASTRA   
49479  2016-04-01 23:55:01                            Tausch_mit_frischen_TÜV   
49496  2016-03-26 13:55:28                                        Bmw_e39_520   
49504  2016-03-13 19:42:57                                 Meriva_Lpg_1_6_16v   
49507  2016-03-16 17:38:32                 BMW_528i_E39_Schaltgetriebe_Unfall   
49525  2016-03-09 17:59:02              Bmw_e36_318_i_Limo_Ringtool_PS_Tuning   
49528  2016-03-28 19:57:41       BMW_E36__316i_Compact_M_Paket_"Schlachtfest"   
49538  2016-03-14 10:55:27                 Golf_3_mit_Tuev_.....PREISWERT!!!!   
49551  2016-03-16 23:40:18        Golf_IV_an_Bastler_zuverkaufen__reserviert_   
49614  2016-03-09 17:49:39  Verkauf_meine__Au

Looking at the data the price does seem reasonable as eBay's marketplace allows you to sell a car with no reserve at auction which is $0.99. We could keep the data for these cheaper cars that are selling at auction, however it will significantly skew the average selling price data. Consequently we will remove these cars.

In [18]:
autos.loc[autos["price"] > 200000] # Cars with a price greater than $200,000

date_crawled                                               name  \
514    2016-03-17 09:53:08                   Ford_Focus_Turnier_1.6_16V_Style   
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
7814   2016-04-04 11:53:31                                        Ferrari_F40   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
12682  2016-03-28 22:48:01  Porsche_GT3_RS__PCCB__Lift___grosser_Exklusiv_...   
14715  2016-03-30 08:37:24                 Rolls_Royce_Phantom_Drophead_Coupe   
22947  2016-03-22 12:54:19                         Bmw_530d_zum_ausschlachten   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
34723  2016-03-23 16:37:29  Porsche_Porsche_911/930_Turbo_3.0__deutsche_Au...   
35923  2016-04-03 07:56:23  Porsche_911_Targa_Exclusive_Edition__1_von_15_...   
36818  2016-03-27 18:37:37                                        Porsche_991   
37585  2016-03-29 11:38:54                                Volkswagen_Jetta_GT   
37840  2016-03-21 10:50:12                                        Porsche_997   
38299  2016-03-28 22:25:25                                Glas_BMW_mit_Wasser   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
43049  2016-03-21 19:53:52                                      2_VW_Busse_T3   
47337  2016-04-05 10:25:38                                    BMW_Z8_roadster   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   
47634  2016-04-04 21:25:21                                        Ferrari_FXX   

       seller offertype     price   abtest vehicletype  registration_year  \
514    privat   Angebot    999999     test       kombi               2009   
2897   privat   Angebot  11111111     test   limousine               1973   
7814   privat   Angebot   1300000  control       coupe               1992   
11137  privat   Angebot  10000000  control       coupe               1960   
12682  privat   Angebot    265000  control       coupe               2016   
14715  privat   Angebot    345000  control      cabrio               2012   
22947  privat   Angebot   1234566  control       kombi               1999   
24384  privat   Angebot  11111111     test         NaN               1995   
27371  privat   Angebot  12345678  control         NaN               2017   
34723  privat   Angebot    299000     test       coupe               1977   
35923  privat   Angebot    295000     test      cabrio               2015   
36818  privat   Angebot    350000  control       coupe               2016   
37585  privat   Angebot    999990     test   limousine               1985   
37840  privat   Angebot    220000     test       coupe               2008   
38299  privat   Angebot    250000     test         NaN               2015   
39377  privat   Angebot  12345678  control         NaN               2018   
39705  privat   Angebot  99999999  control   limousine               1999   
42221  privat   Angebot  27322222  control   limousine               2014   
43049  privat   Angebot    999999     test         bus               1981   
47337  privat   Angebot    259000     test      cabrio               2001   
47598  privat   Angebot  12345678  control   limousine               2001   
47634  privat   Angebot   3890000     test       coupe               2006   

         gearbox  powerps        model  odometer_km  registration_month  \
514      manuell      101        focus       125000                   4   
2897     manuell       48       escort        50000                   3   
7814         NaN        0          NaN        50000                  12   
11137    manuell      368          NaN       100000                   1   
126

Many cars advertised in this price range are legitimate. Generally, those cars over $350,000 are priced incorrectly so we will use this as a guide to remove the outliers.

#### PowerPS Outliers

In [19]:
autos.loc[autos["powerps"] <18] # Cars with less than 18 PS

date_crawled                                               name  \
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
8      2016-03-22 16:51:34                                         Seat_Arosa   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
32     2016-03-20 05:03:03                               Corsa_mit_TÜV_5.2016   
35     2016-03-29 21:52:56                                        Ford_Mondeo   
46     2016-03-31 10:53:28                                       BMW_mit__Lpg   
52     2016-03-25 18:50:03  Senator_A_3.0E_Karosserie_restauriert_m._viele...   
59     2016-03-17 17:50:54               Mercedes_A_Klasse_W_168__A_140_gruen   
64     2016-04-05 07:36:19                 Autotransport__Abschlepp_Schlepper   
71     2016-03-28 19:39:35  Suche_Opel_Astra_F__Corsa_oder_Kadett_E_mit_Re...   
74     2016-03-06 02:00:38                                         Opel_Kombi   
85     2016-03-11 12:36:21                                  Toyota_yaris_cool   
87     2016-03-29 23:37:22                      Bmw_520_e39_zum_ausschlachten   
92     2016-03-25 19:47:32                                       Zu_verkaufen   
99     2016-04-05 09:48:54                    Peugeot_207_CC___Cabrio_Bj_2011   
101    2016-03-22 17:51:49    Schnaepchen_in_einem_Jahr_OLDTEIMER_KENNZEICHEN   
102    2016-03-22 11:57:49                                Ford_Ka_dunkel_blau   
124    2016-03-16 14:51:35                     Passat_2.0_TDI_DPF_Comfortline   
139    2016-03-13 19:43:32                                Mercedes_Benz_E_320   
144    2016-03-31 15:52:58  Skoda_Felicia_8_fach_bereift_86300_km___kein_Rost   
155    2016-03-07 18:45:47                                  VW_Touran__140_PS   
164    2016-03-13 20:39:16  Opel_Meriva__nur_76000_Km__unfallfrei__scheckh...   
167    2016-04-02 19:43:45  Suche_VW_Multivan_Innenausstattung_Set_oder_TE...   
175    2016-03-19 15:48:21                                          MG_MGB_GT   
181    2016-03-30 23:45:05       Peugeot_206_Klimaanlage_65kw_TÜV_bis_01.2017   
191    2016-03-07 05:39:55               Ford_Mondeo_Kombi_Diesel_ca.207000km   
192    2016-03-12 22:40:28                                        Opel_VIVARO   
199    2016-03-12 08:54:25                           Renault_clio_TÜV_09_2017   
210    2016-04-03 21:00:02                KIA_Carnival_Automatik_Motor_defekt   
...                    ...                                                ...   
49741  2016-04-01 16:44:42                                   Verkaufe_Ford_Ka   
49774  2016-03-08 14:51:37                                320_d_Edition_Sport   
49777  2016-03-12 13:52:37                       A4_S_line_Quattro_FULL_PAKET   
49798  2016-03-21 02:00:39                                 Opel_Astra_Caravan   
49806  2016-03-15 17:56:40                               Mitsubishi__Colt_1_1   
49815  2016-03-08 10:06:22  SUCHE_TIPPS___Ford_Mustang_Shelby_GT_350_500_K...   
49817  2016-03-16 17:58:35                                 Mercedes_E240_211k   
49842  2016-04-02 14:42:45                 Mercedes_Benz_Vito_111_CDI_Kompakt   
49844  2016-03-26 19:43:51                                Fahrzeug_Ankauf!!!!   
49849  2016-04-05 11:50:41                  1997_Opel_Vectra_B_1.8_CD_Mit_TÜV   
49860  2016-04-03 14:49:24                           Tausche_gegen_Opel_Agila   
49865  2016-03-10 01:37:42  Mercedes_CLK_270_CDI_TÜV_bis_11_2017_klimaauto...   
49868  2016-03-26 21:50:00                            C_Klasse_200_Kompressor   
49870  2016-03-08 18:50:51         A2_1.4_16V_S_Line_Leder_interieur_Panorama   
49871  2016-03-12 19:00:30                       Ford_Fiesta_1_4_TÜV_bis_2018   
49874  2016-04-02 18:25:23                              Ford_KA_mit_Tuev_2017   
49876  2016-03-22 17:57:24                             Audi_a5_3.0_tdi_s_line   
49892  2016-03-14 14:57:03                   

When browsing the powerps values, to the point of 18ps, many advertisements seem legitimate. Hence we will remove any cars with less than 18ps. Lets look at cars with more than 550PS

In [20]:
autos.loc[autos["powerps"] > 550] # Cars with more than 550PS

date_crawled                                               name  \
160    2016-03-31 02:36:24                                         BMW_M5_DKG   
1699   2016-04-04 19:49:19                        Opel_Corsa_1.0_Motor_ecotek   
2220   2016-03-30 17:56:27                                Ford_ka_top_zustand   
2670   2016-04-02 15:47:00                              Verkaufe_Ford_Focus_!   
2876   2016-03-19 06:36:23                 Golf_3_Cabrio_voll_fahrbereit_tuev   
3753   2016-04-03 18:47:14                                         VW_Polo_9n   
4279   2016-03-29 19:51:19                        Citroen_C4_1.6Hdi__94.500km   
4405   2016-03-21 19:53:24                             VW_Golf_Automatik_Grau   
4464   2016-03-29 16:47:47            Zu_verkaufen_Mercedes_A_160_mit_neu_TÜV   
4777   2016-04-03 12:45:25              Audi_tt_bj_2000_Unfall.._laeuft_top..   
5342   2016-03-08 17:36:57                  Mercedes_Benz_CL_65_AMG_Automatik   
6923   2016-03-06 22:47:40              Verkaufe_einen_schoenen_Stadtflitzer.   
7556   2016-03-23 23:56:04         Kaufe_alle_A4_B5_egal_in_welchem_zustand!!   
7798   2016-03-15 14:36:47    Mercedes_Benz_C_63_AMG_Coupe_AMG_SPEEDSHIFT_MCT   
8000   2016-03-16 00:18:43                                        Opel_zafira   
8446   2016-03-24 05:03:36              Bentley_Continental_Flying_Spur_Speed   
9335   2016-03-20 17:53:53                     Golf_3_Vr6_Turbo_4Motion_640PS   
9968   2016-03-14 15:50:17  Start_Stop_Leder_Lenkrad_mit_neuen_Alweterrreifen   
10454  2016-03-09 23:54:51       Audi_RS6_Avant_850PS_Getriebe_NEU_21"_Felgen   
10556  2016-04-01 06:02:10                                         UNFAL_Auto   
10659  2016-03-18 11:47:40                            VW_POLO_BASTLERFAHRZEUG   
10668  2016-03-29 18:46:51                                           BMW_X5_M   
10813  2016-03-12 09:58:22        Volkswagen_Kaefer_1303_1972_von_2e_eigenere   
11009  2016-03-15 21:55:11               Smart_fortwo_Silver_**Brabus_Style**   
11311  2016-04-01 02:36:43         Vw__zu__verkaufen__Caravelle_Confort_kurz_   
11433  2016-03-30 21:44:21  Audi_R8_V10_plus_5.2_FSI_quattro_S_tronic_Akra...   
11893  2016-03-10 11:53:22    Original__Lamborgini_Diablo_Schutzhaube_91_2002   
11914  2016-03-09 22:56:46                 Volkswagen_Caravelle_VW_T4_2_5_TDI   
12130  2016-03-17 19:57:47                                     Audi_RS6_Avant   
13573  2016-03-12 10:56:09         A_140__Bastler_fahrzeug__mit_TÜV_bis_08/17   
...                    ...                                                ...   
34634  2016-03-08 19:57:01               Chevrolet_Camaro_ZL1_Coupe_Automatic   
35039  2016-03-10 08:57:25                               Top_Cabrio_A4_s_line   
36191  2016-03-25 14:55:04                            1.HAND__wenig_Kilometer   
36421  2016-03-20 16:51:48                           Verkaufe_meinen_bmw_525d   
36973  2016-03-14 13:58:16   Mercedes_E_Klasse_280_CDI_sehr_gepflegt_67500_km   
37651  2016-03-29 19:44:48                           verkauf__bastler__export   
39221  2016-03-27 11:52:47                               Seat_Marbella_kultig   
40017  2016-03-05 17:48:02                              Ford_Focus_1.8_Diesel   
40091  2016-03-14 13:58:10                                         Ford_Focus   
40096  2016-03-09 18:43:44                               Ford_Streetka_1.6_8V   
40625  2016-03-24 11:51:57                  VW_TOURAN_HIGHLINE_2.0_TDI_DIESEL   
40693  2016-03-08 11:50:02                                     Audi_RS6_Avant   
40807  2016-03-17 17:56:36  Mercedes_Benz_S_4Matic_*Carpristo*Bang&Olufsen...   
40975  2016-03-09 10:54:49                    Bentley_Continental_Flying_Spur   
41172  2016-04-03 14:45:36                      Suzuki_Baleno_1_3l_GS_TÜV_neu   
41673  2016-04-02 11:51:51                                Opel_Agila_Silber_?   
43870  2016-03-08 03:02:24  VW_Passat_2_0_TDi_Inzahlungn._/Tausche_gegen_T...   
43987  2016-03-31 03:58:00         BMW_M5_DKG

The majority of cars in this power range have incorrect power levels. We will use this as the upper bounds to remove outliers.

### Removing the Outliers from Pricing, PowerPS and Registration Year Data

In [21]:
'''
Removing the outliers from the pricing data
'''

autos = autos[autos["price"].between(100,350000)]

In [22]:
autos["price"].describe()

count     48224.000000
mean       5930.371433
std        9078.372762
min         100.000000
25%        1250.000000
50%        3000.000000
75%        7499.000000
max      350000.000000
Name: price, dtype: float64

The values in the price column are now manageable for data analysis.

After that exercise we removed 1,776 rows.

In [23]:
autos["registration_year"].describe()

count    48224.000000
mean      2004.730964
std         87.897388
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The registration year column has a significant range of values ranging from 1000 to 9999. These are certainly some odd values. As the listings end in 2016, any car from 2016 onwards should be removed while any car registered before 1900 should also be removed.

In [24]:
autos = autos[autos["registration_year"].between(1900,2016)]

In [25]:
autos["registration_year"].describe()

count    46352.000000
mean      2002.939787
std          7.127146
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

The registration data looks much better now.

In [26]:
autos["registration_year"].value_counts(normalize=True).sort_index(ascending=False)

2016    0.025932
2015    0.008198
2014    0.014282
2013    0.017281
2012    0.028219
2011    0.034907
2010    0.034238
2009    0.044874
2008    0.047679
2007    0.049038
2006    0.057560
2005    0.062802
2004    0.058228
2003    0.058099
2002    0.053439
2001    0.056718
2000    0.066966
1999    0.062112
1998    0.050483
1997    0.041530
1996    0.029233
1995    0.025738
1994    0.013505
1993    0.009061
1992    0.007918
1991    0.007292
1990    0.007163
1989    0.003689
1988    0.002869
1987    0.001553
          ...   
1968    0.000561
1967    0.000561
1966    0.000475
1965    0.000367
1964    0.000259
1963    0.000173
1962    0.000086
1961    0.000129
1960    0.000475
1959    0.000129
1958    0.000086
1957    0.000043
1956    0.000086
1955    0.000043
1954    0.000043
1953    0.000022
1952    0.000022
1951    0.000043
1950    0.000022
1948    0.000022
1943    0.000022
1941    0.000043
1939    0.000022
1938    0.000022
1937    0.000086
1934    0.000043
1931    0.000022
1929    0.0000

In [27]:
autos = autos[autos["powerps"].between(18,550)] #Removing Outliers from PowerPS

In [28]:
autos["powerps"].describe()

count    41886.000000
mean       126.493554
std         60.990597
min         18.000000
25%         80.000000
50%        116.000000
75%        150.000000
max        544.000000
Name: powerps, dtype: float64

These figures for Powerps look much better now.

### Removing Unnecessary Columns

After removing the outliers, we will now remove unnecessary columns for our analysis.

In [29]:
autos.info() #To check what columns we currently have.

<class 'pandas.core.frame.DataFrame'>
Int64Index: 41886 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          41886 non-null object
name                  41886 non-null object
seller                41886 non-null object
offertype             41886 non-null object
price                 41886 non-null int64
abtest                41886 non-null object
vehicletype           40437 non-null object
registration_year     41886 non-null int64
gearbox               41185 non-null object
powerps               41886 non-null int64
model                 40321 non-null object
odometer_km           41886 non-null int64
registration_month    41886 non-null int64
fueltype              39769 non-null object
brand                 41886 non-null object
unrepaired_damage     35991 non-null object
ad_created            41886 non-null object
nrofpictures          41886 non-null int64
postalcode            41886 non-null int64
last_seen             41886 non-null object
dtypes: int64(7), 

In [30]:
autos.drop(['seller', 'offertype', 'nrofpictures', 'date_crawled', 'last_seen', 'ad_created'], axis='columns', inplace=True)

In [31]:
autos.info() #See if the redundant columns have now been removed.

<class 'pandas.core.frame.DataFrame'>
Int64Index: 41886 entries, 0 to 49999
Data columns (total 14 columns):
name                  41886 non-null object
price                 41886 non-null int64
abtest                41886 non-null object
vehicletype           40437 non-null object
registration_year     41886 non-null int64
gearbox               41185 non-null object
powerps               41886 non-null int64
model                 40321 non-null object
odometer_km           41886 non-null int64
registration_month    41886 non-null int64
fueltype              39769 non-null object
brand                 41886 non-null object
unrepaired_damage     35991 non-null object
postalcode            41886 non-null int64
dtypes: int64(6), object(8)
memory usage: 4.8+ MB


### Filling in the Blanks

From the last autos.info() run, we can see that there are a few columns with values that are blank. For example, unrepaired_damage has 39,769 values while date_crawled has 41,886. 

We can either remove all of these rows with missing values, or we can replace them with "N/A". In my opinion, we're probably better off keeping the rows and using "N/A" to fill the gaps as there is still usable data here.

Of concern the fueltype and unrepaired_damage columns will require filling in.

In [32]:
#Creating some filters of rows with columns containing blanks to explore the data

bool_series_ft = pd.isnull(autos["fueltype"])
bool_series_unrepaired = pd.isnull(autos["unrepaired_damage"])

In [33]:
#Viewing rows with blanks in the unrepaired_damage column
autos[bool_series_unrepaired]


name  price   abtest  \
5      Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   7900     test   
6      VW_Golf_III_GT_Special_Electronic_Green_Metall...    300     test   
14     Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   3999     test   
18                  Verkaufen_mein_bmw_e36_320_i_touring    300  control   
20     Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   3500     test   
34                               Bmw_530d__navi_tuv03/18   2600     test   
42     Vw_Polo_l.0_/60_PS_Blue_Motion_Technologie_Son...  11900  control   
56                                          Renault_Clio    650     test   
67                                      Jeep_Wrangler_YJ   4800     test   
73                           VW_T3_Doka_1_7D_TÜV_07/2017   2700     test   
76                            BMW_318i_neustes_Model_0Km  31999  control   
77                                  Volvo_V40___Tuev_neu   2200     test   
79                                Seat_Ibiza_1.9_TDI__Fr   6300  control   
98                     Audi_A4_Cabriolet_2.0_TFSI_S_Line  13500     test   
107                           zu_verkaufen_Fat_punto_1.9    550     test   
121                       verkaufe_Ford_focus_1_8_Diesel   1100  control   
135    Opel_Meriva_B_Panoramadach__Sitz__und_Lenkradh...   8500  control   
138    Ford_C_Max_Titanium_1_0L._125Ps_Benziner__Magn...  18300     test   
143    Mercedes_Benz_E_Klasse_W211_320_CDI_Vollaussta...   9655     test   
147                            VW_Polo_Ersatzteilspender    600  control   
169    VW_Golf_Variant_1.6_SR__Klima__Schiebedach__AH...   1700  control   
174                                  Fiat_Punto_sporting    300     test   
187                               VW_Polo_Kombi_TÜV_2017    900  control   
190                               An_Bastler_oder_Export    800     test   
203                                     BMW_316i_Touring    750     test   
217                                PORSCHE_924S__11/1989   8900  control   
221                        Opel_Corsa_B_Bastler_Kein_TÜV    150     test   
222                                    Opel_Astra_Cabrio   1600  control   
231                      VW_Passat_35i__2.0Liter__115_PS    350  control   
239                              Mercedes_benz_c_220_cdi   3500  control   
...                                                  ...    ...      ...   
49729                         WV_Touran_Highline_2.0_Tdi   3050     test   
49730                              Volkswagen_Sharan_2.0    999     test   
49732                      Volkswagen_Golf_Cabriolet_TDI   2800     test   
49742                                     Renault_Twingo   3490  control   
49758                             Opel_astra_1_6_eco_tec    600  control   
49759                       Ford_Focus_2.0_TDCi_Titanium   3600     test   
49763                        318ti_compact_M_ausstattung   2100  control   
49771  BMW_120i_150ps_Lederausstattung_Klima_Multi_Le...  11850  control   
49784            2x_Golf_I_Cabrio_Bel_Air__auch_einzeln_   2800     test   
49785                                    Bmw_E30_335_m30   6500     test   
49795  Tausche_7___Sitzer_mit_190_PS___Automatik____>...   3999  control   
49837                               Ford_Focus_Kombi_1_8   2000  control   
49839                       Mercedes_Benz_320_Avantgarde   1999     test   
49852                                   TOP__Golf_3_1.8l   1450  control   
49862                               opel_astra_F_caravan   1100  control   
49875               Golf_5_Variant_ab_April_zu_verkaufen   7900  control   
49879  Tausche_meinen__Daimlerchrysler_Stratus_Cabrio...   2500     test   
49889             Seat_Ibizia_Stella_1.4__60_PS__Bj.2000   1350  control   
49890                                      Nissan_Almira    990     test   
49904  Susuki_Baleno_TÜV_3.2018_Optisch_und_technisch...   1000     test   
49920                                        Mazda_323_F    700     test   
49921  Smart_ForTwo_Cabr

From the above, we can see that for the majority of rows missing data in the unrepaired_damage column, the data is still good so we will replace the missing values with 'N/A'.

In [34]:
autos["unrepaired_damage"].fillna('N/A',inplace=True) #Replace the blank values.

In [35]:
autos["unrepaired_damage"].value_counts() #Check to see that this is ok now.

nein    32022
N/A      5895
ja       3969
Name: unrepaired_damage, dtype: int64

In [36]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 41886 entries, 0 to 49999
Data columns (total 14 columns):
name                  41886 non-null object
price                 41886 non-null int64
abtest                41886 non-null object
vehicletype           40437 non-null object
registration_year     41886 non-null int64
gearbox               41185 non-null object
powerps               41886 non-null int64
model                 40321 non-null object
odometer_km           41886 non-null int64
registration_month    41886 non-null int64
fueltype              39769 non-null object
brand                 41886 non-null object
unrepaired_damage     41886 non-null object
postalcode            41886 non-null int64
dtypes: int64(6), object(8)
memory usage: 4.8+ MB


In [37]:
autos["fueltype"].fillna('N/A',inplace=True) #Replace blanks in fueltype column

In [38]:
autos["fueltype"].value_counts()

benzin     26005
diesel     13041
N/A         2117
lpg          605
cng           65
hybrid        37
elektro        8
andere         8
Name: fueltype, dtype: int64

In [39]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 41886 entries, 0 to 49999
Data columns (total 14 columns):
name                  41886 non-null object
price                 41886 non-null int64
abtest                41886 non-null object
vehicletype           40437 non-null object
registration_year     41886 non-null int64
gearbox               41185 non-null object
powerps               41886 non-null int64
model                 40321 non-null object
odometer_km           41886 non-null int64
registration_month    41886 non-null int64
fueltype              41886 non-null object
brand                 41886 non-null object
unrepaired_damage     41886 non-null object
postalcode            41886 non-null int64
dtypes: int64(6), object(8)
memory usage: 4.8+ MB


### Removing Remaining Rows with Blanks

Now that the data is relatively clean, for completeness we will now remove all rows that contain blanks.

In [40]:
cleaned_autos = autos.dropna()

In [41]:
cleaned_autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 38577 entries, 0 to 49999
Data columns (total 14 columns):
name                  38577 non-null object
price                 38577 non-null int64
abtest                38577 non-null object
vehicletype           38577 non-null object
registration_year     38577 non-null int64
gearbox               38577 non-null object
powerps               38577 non-null int64
model                 38577 non-null object
odometer_km           38577 non-null int64
registration_month    38577 non-null int64
fueltype              38577 non-null object
brand                 38577 non-null object
unrepaired_damage     38577 non-null object
postalcode            38577 non-null int64
dtypes: int64(6), object(8)
memory usage: 4.4+ MB


We now have 38,577 rows remaining that we can use for our analysis.

## Analysis

### Brand Analysis by Average Price

In [42]:
'''
We want to select the top 20 brands by ad-listing numbers
'''

filtered_brands = cleaned_autos["brand"].value_counts().head(20)
print(filtered_brands)

volkswagen       8145
bmw              4503
opel             4026
mercedes_benz    3856
audi             3571
ford             2627
renault          1739
peugeot          1133
fiat              939
seat              728
skoda             695
mazda             585
nissan            584
citroen           536
toyota            508
smart             480
hyundai           397
mini              378
volvo             371
mitsubishi        315
Name: brand, dtype: int64


In [43]:
filtered_brands = filtered_brands.index
print(filtered_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'nissan', 'citroen',
       'toyota', 'smart', 'hyundai', 'mini', 'volvo', 'mitsubishi'],
      dtype='object')


In [44]:
avg_price_by_brand = {}

for brands in filtered_brands:
    selected_rows = cleaned_autos[cleaned_autos["brand"] == brands]
    mean = selected_rows["price"].mean()
    avg_price_by_brand[brands] = mean

In [45]:
avg_price_by_brand

{'audi': 9958.29095491459,
 'bmw': 8846.654452587165,
 'citroen': 4038.001865671642,
 'fiat': 3057.6900958466454,
 'ford': 4097.900647125999,
 'hyundai': 5782.153652392947,
 'mazda': 4594.270085470085,
 'mercedes_benz': 9158.69917012448,
 'mini': 10773.354497354498,
 'mitsubishi': 3530.146031746032,
 'nissan': 5209.66095890411,
 'opel': 3282.6915052160953,
 'peugeot': 3378.1932921447483,
 'renault': 2738.608395629672,
 'seat': 4825.070054945055,
 'skoda': 6759.576978417266,
 'smart': 3881.6041666666665,
 'toyota': 5407.854330708661,
 'volkswagen': 5950.0902394106815,
 'volvo': 5326.455525606469}

Previously we saw that Volkswagen, BMW, Opel, Mercedes Benz and Audi made up the top 5 number of ad-listings by volume. 

The average price by brand gives us some insight into the German car market. 

French and Italian cars such as Peugeot, Renault, Fiat & Citroen are generally at the more affordable end of the spectrum while German brands tend to be at the higher end of the price range. 

Sandwiched in the middle are the Japanese and Korean brands with some of the other minor European brands. 

In [46]:
avg_mileage_by_brand = {}

for brands in filtered_brands:
    selected_rows = cleaned_autos[cleaned_autos["brand"] == brands]
    mean = selected_rows["odometer_km"].mean()
    avg_mileage_by_brand[brands] = mean
    

In [47]:
avg_mileage_by_brand

{'audi': 128011.76141136937,
 'bmw': 132164.11281367976,
 'citroen': 117957.0895522388,
 'fiat': 115969.11608093716,
 'ford': 123587.74267224972,
 'hyundai': 103438.28715365239,
 'mazda': 123504.2735042735,
 'mercedes_benz': 130282.67634854771,
 'mini': 88095.23809523809,
 'mitsubishi': 125746.03174603175,
 'nissan': 116515.4109589041,
 'opel': 128669.8956780924,
 'peugeot': 125701.67696381289,
 'renault': 127032.77745830937,
 'seat': 120151.0989010989,
 'skoda': 109510.79136690647,
 'smart': 97333.33333333333,
 'toyota': 115187.00787401575,
 'volkswagen': 127653.7753222836,
 'volvo': 138450.13477088948}

In [48]:
bmp_series = pd.Series(avg_price_by_brand)
print(bmp_series)

audi              9958.290955
bmw               8846.654453
citroen           4038.001866
fiat              3057.690096
ford              4097.900647
hyundai           5782.153652
mazda             4594.270085
mercedes_benz     9158.699170
mini             10773.354497
mitsubishi        3530.146032
nissan            5209.660959
opel              3282.691505
peugeot           3378.193292
renault           2738.608396
seat              4825.070055
skoda             6759.576978
smart             3881.604167
toyota            5407.854331
volkswagen        5950.090239
volvo             5326.455526
dtype: float64


In [49]:
df = pd.DataFrame(bmp_series, columns=['mean_price'])
df

mean_price
audi            9958.290955
bmw             8846.654453
citroen         4038.001866
fiat            3057.690096
ford            4097.900647
hyundai         5782.153652
mazda           4594.270085
mercedes_benz   9158.699170
mini           10773.354497
mitsubishi      3530.146032
nissan          5209.660959
opel            3282.691505
peugeot         3378.193292
renault         2738.608396
seat            4825.070055
skoda           6759.576978
smart           3881.604167
toyota          5407.854331
volkswagen      5950.090239
volvo           5326.455526

In [50]:
bmm_series = pd.Series(avg_mileage_by_brand)
print(bmm_series)

audi             128011.761411
bmw              132164.112814
citroen          117957.089552
fiat             115969.116081
ford             123587.742672
hyundai          103438.287154
mazda            123504.273504
mercedes_benz    130282.676349
mini              88095.238095
mitsubishi       125746.031746
nissan           116515.410959
opel             128669.895678
peugeot          125701.676964
renault          127032.777458
seat             120151.098901
skoda            109510.791367
smart             97333.333333
toyota           115187.007874
volkswagen       127653.775322
volvo            138450.134771
dtype: float64


In [51]:
df["mean_mileage"] = bmm_series
df

mean_price   mean_mileage
audi            9958.290955  128011.761411
bmw             8846.654453  132164.112814
citroen         4038.001866  117957.089552
fiat            3057.690096  115969.116081
ford            4097.900647  123587.742672
hyundai         5782.153652  103438.287154
mazda           4594.270085  123504.273504
mercedes_benz   9158.699170  130282.676349
mini           10773.354497   88095.238095
mitsubishi      3530.146032  125746.031746
nissan          5209.660959  116515.410959
opel            3282.691505  128669.895678
peugeot         3378.193292  125701.676964
renault         2738.608396  127032.777458
seat            4825.070055  120151.098901
skoda           6759.576978  109510.791367
smart           3881.604167   97333.333333
toyota          5407.854331  115187.007874
volkswagen      5950.090239  127653.775322
volvo           5326.455526  138450.134771

Comparing the average price and average mileage, we can see that for cars in a similar group, the higher the average mileage, the lower the average price. Take for example the French car makers Renault, Citroen and Peugeot. These three car makers are competitive in retail price points and car models. In fact, Citroen and Peugeot are basically identical as they share platforms and are the one company. Comparing the average mileage for Citroen and Peugeot, we see that the lower the mileage, the higher the average price. 

The same can be said for the big 3 German luxury car makers, BMW, Audi and Mercedes-Benz too as well as grouping the Japanese and Korean brands together. We could make a reasonable assumption that mileage has a negative relationship with price.

### Most Popular Models by Brand

To find out which is the most listed model by brand we will use a similar technique to the previous analysis.

In [90]:
most_listed_model_by_brand = {}

for brands in filtered_brands:
    selected_rows = cleaned_autos["brand"] == brands #for each unique brand
    sorted_rows = cleaned_autos[selected_rows] #use this as a filter
    d_rows = sorted_rows["model"].value_counts() # find the value counts of each model
    most_listed_model_by_brand[brands] = d_rows.index[0] # return the most popular model from the value counts

In [89]:
most_listed_model_by_brand

{'audi': 'a4',
 'bmw': '3er',
 'citroen': 'andere',
 'fiat': 'punto',
 'ford': 'focus',
 'hyundai': 'i_reihe',
 'mazda': '6_reihe',
 'mercedes_benz': 'c_klasse',
 'mini': 'cooper',
 'mitsubishi': 'colt',
 'nissan': 'micra',
 'opel': 'corsa',
 'peugeot': '2_reihe',
 'renault': 'twingo',
 'seat': 'ibiza',
 'skoda': 'octavia',
 'smart': 'fortwo',
 'toyota': 'yaris',
 'volkswagen': 'golf',
 'volvo': 'v70'}

In [73]:
'''
Code for the above loop to return the top value counts for each model by brand.
'''

#b_bool = cleaned_autos["brand"] == "bmw"
#c_bool = cleaned_autos[b_bool]
#d_bool = c_bool["model"].value_counts()
#d_bool

3er        2363
5er        1050
1er         490
x_reihe     277
7er         116
z_reihe     112
m_reihe      39
andere       29
6er          27
Name: model, dtype: int64

In [91]:
pop_model_series = pd.Series(most_listed_model_by_brand)
print(pop_model_series)

audi                   a4
bmw                   3er
citroen            andere
fiat                punto
ford                focus
hyundai           i_reihe
mazda             6_reihe
mercedes_benz    c_klasse
mini               cooper
mitsubishi           colt
nissan              micra
opel                corsa
peugeot           2_reihe
renault            twingo
seat                ibiza
skoda             octavia
smart              fortwo
toyota              yaris
volkswagen           golf
volvo                 v70
dtype: object


In [93]:
df1 = pd.DataFrame(pop_model_series, columns=['top_model'])
df1

top_model
audi                 a4
bmw                 3er
citroen          andere
fiat              punto
ford              focus
hyundai         i_reihe
mazda           6_reihe
mercedes_benz  c_klasse
mini             cooper
mitsubishi         colt
nissan            micra
opel              corsa
peugeot         2_reihe
renault          twingo
seat              ibiza
skoda           octavia
smart            fortwo
toyota            yaris
volkswagen         golf
volvo               v70

In [97]:
most_listed_model_by_brand_no = {} #Number of listings for the most popular model

for brands in filtered_brands:
    selected_rows = cleaned_autos["brand"] == brands #for each unique brand
    sorted_rows = cleaned_autos[selected_rows] #use this as a filter
    d_rows = sorted_rows["model"].value_counts() # find the value counts of each model
    most_listed_model_by_brand_no[brands] = d_rows[0] # return the number of listings for the most popular model from the value counts

In [98]:
most_listed_model_by_brand_no

{'audi': 1128,
 'bmw': 2363,
 'citroen': 179,
 'fiat': 331,
 'ford': 659,
 'hyundai': 166,
 'mazda': 144,
 'mercedes_benz': 1033,
 'mini': 251,
 'mitsubishi': 98,
 'nissan': 194,
 'opel': 1292,
 'peugeot': 540,
 'renault': 470,
 'seat': 294,
 'skoda': 286,
 'smart': 410,
 'toyota': 113,
 'volkswagen': 3176,
 'volvo': 81}

In [99]:
pop_model_series_no = pd.Series(most_listed_model_by_brand_no)
print(pop_model_series_no)

audi             1128
bmw              2363
citroen           179
fiat              331
ford              659
hyundai           166
mazda             144
mercedes_benz    1033
mini              251
mitsubishi         98
nissan            194
opel             1292
peugeot           540
renault           470
seat              294
skoda             286
smart             410
toyota            113
volkswagen       3176
volvo              81
dtype: int64


In [100]:
df1["no_of_listings"] = pop_model_series_no
df1

top_model  no_of_listings
audi                 a4            1128
bmw                 3er            2363
citroen          andere             179
fiat              punto             331
ford              focus             659
hyundai         i_reihe             166
mazda           6_reihe             144
mercedes_benz  c_klasse            1033
mini             cooper             251
mitsubishi         colt              98
nissan            micra             194
opel              corsa            1292
peugeot         2_reihe             540
renault          twingo             470
seat              ibiza             294
skoda           octavia             286
smart            fortwo             410
toyota            yaris             113
volkswagen         golf            3176
volvo               v70              81

From the above, the Volkswagen Golf is the most listed model from the sample data. With the exception of the Volvo V70, the most listed cars from each brand are generally small to medium cars.

### Price Difference for Damaged Cars

In [127]:
avg_damaged_by_brand_ja = {}

for brands in filtered_brands:
    selected_rows = cleaned_autos[cleaned_autos["brand"] == brands]
    damage_filter = selected_rows[selected_rows["unrepaired_damage"] == "ja"]
    mean = damage_filter["price"].mean()
    avg_damaged_by_brand_ja[brands] = mean

In [136]:
avg_damaged_by_brand_ja

{'audi': 3398.9103448275864,
 'bmw': 3823.460641399417,
 'citroen': 1887.3035714285713,
 'fiat': 1259.58,
 'ford': 1574.2539682539682,
 'hyundai': 2836.860465116279,
 'mazda': 1414.0144927536232,
 'mercedes_benz': 4378.076923076923,
 'mini': 4749.090909090909,
 'mitsubishi': 1485.953488372093,
 'nissan': 2066.9875,
 'opel': 1535.2227272727273,
 'peugeot': 1476.8731343283582,
 'renault': 1261.066037735849,
 'seat': 1930.7142857142858,
 'skoda': 3839.5737704918033,
 'smart': 1558.7692307692307,
 'toyota': 3721.9333333333334,
 'volkswagen': 2376.1651234567903,
 'volvo': 1893.15}

In [137]:
damaged_cars_series = pd.Series(avg_damaged_by_brand_ja)
print(damaged_cars_series)

audi             3398.910345
bmw              3823.460641
citroen          1887.303571
fiat             1259.580000
ford             1574.253968
hyundai          2836.860465
mazda            1414.014493
mercedes_benz    4378.076923
mini             4749.090909
mitsubishi       1485.953488
nissan           2066.987500
opel             1535.222727
peugeot          1476.873134
renault          1261.066038
seat             1930.714286
skoda            3839.573770
smart            1558.769231
toyota           3721.933333
volkswagen       2376.165123
volvo            1893.150000
dtype: float64


In [138]:
df2 = pd.DataFrame(damaged_cars_series, columns=['damaged_cars'])
df2

damaged_cars
audi            3398.910345
bmw             3823.460641
citroen         1887.303571
fiat            1259.580000
ford            1574.253968
hyundai         2836.860465
mazda           1414.014493
mercedes_benz   4378.076923
mini            4749.090909
mitsubishi      1485.953488
nissan          2066.987500
opel            1535.222727
peugeot         1476.873134
renault         1261.066038
seat            1930.714286
skoda           3839.573770
smart           1558.769231
toyota          3721.933333
volkswagen      2376.165123
volvo           1893.150000

In [129]:
avg_damaged_by_brand_nein = {}

for brands in filtered_brands:
    selected_rows = cleaned_autos[cleaned_autos["brand"] == brands]
    damage_filter = selected_rows[selected_rows["unrepaired_damage"] == "nein"]
    mean = damage_filter["price"].mean()
    avg_damaged_by_brand_nein[brands] = mean

In [139]:
avg_damaged_by_brand_nein

{'audi': 11228.818658280923,
 'bmw': 9695.068355119825,
 'citroen': 4424.62945368171,
 'fiat': 3548.212250712251,
 'ford': 4856.3175572519085,
 'hyundai': 6297.921875,
 'mazda': 5394.134831460674,
 'mercedes_benz': 10008.875866414619,
 'mini': 11141.760683760684,
 'mitsubishi': 4197.515021459228,
 'nissan': 6027.091517857143,
 'opel': 3787.832114361702,
 'peugeot': 3847.3859447004606,
 'renault': 3216.0781860828774,
 'seat': 5399.3694158075605,
 'skoda': 7266.2626086956525,
 'smart': 4068.8693467336684,
 'toyota': 5628.972911963882,
 'volkswagen': 6743.192043681747,
 'volvo': 6020.924092409241}

In [140]:
undamaged_cars_series = pd.Series(avg_damaged_by_brand_nein)
print(undamaged_cars_series)

audi             11228.818658
bmw               9695.068355
citroen           4424.629454
fiat              3548.212251
ford              4856.317557
hyundai           6297.921875
mazda             5394.134831
mercedes_benz    10008.875866
mini             11141.760684
mitsubishi        4197.515021
nissan            6027.091518
opel              3787.832114
peugeot           3847.385945
renault           3216.078186
seat              5399.369416
skoda             7266.262609
smart             4068.869347
toyota            5628.972912
volkswagen        6743.192044
volvo             6020.924092
dtype: float64


In [141]:
df2["undamaged_cars"] = undamaged_cars_series
df2

damaged_cars  undamaged_cars
audi            3398.910345    11228.818658
bmw             3823.460641     9695.068355
citroen         1887.303571     4424.629454
fiat            1259.580000     3548.212251
ford            1574.253968     4856.317557
hyundai         2836.860465     6297.921875
mazda           1414.014493     5394.134831
mercedes_benz   4378.076923    10008.875866
mini            4749.090909    11141.760684
mitsubishi      1485.953488     4197.515021
nissan          2066.987500     6027.091518
opel            1535.222727     3787.832114
peugeot         1476.873134     3847.385945
renault         1261.066038     3216.078186
seat            1930.714286     5399.369416
skoda           3839.573770     7266.262609
smart           1558.769231     4068.869347
toyota          3721.933333     5628.972912
volkswagen      2376.165123     6743.192044
volvo           1893.150000     6020.924092

In [134]:
avg_damaged_by_brand_na = {}

for brands in filtered_brands:
    selected_rows = cleaned_autos[cleaned_autos["brand"] == brands]
    damage_filter = selected_rows[selected_rows["unrepaired_damage"] == "N/A"]
    mean = damage_filter["price"].mean()
    avg_damaged_by_brand_na[brands] = mean

In [135]:
avg_damaged_by_brand_na

{'audi': 5819.7947494033415,
 'bmw': 5993.334016393443,
 'citroen': 3320.5254237288136,
 'fiat': 1856.7007299270074,
 'ford': 2094.037463976945,
 'hyundai': 4652.794117647059,
 'mazda': 2671.7042253521126,
 'mercedes_benz': 5751.973105134474,
 'mini': 6833.125,
 'mitsubishi': 1796.8974358974358,
 'nissan': 3159.75,
 'opel': 1984.1159169550174,
 'peugeot': 2214.206106870229,
 'renault': 1539.233870967742,
 'seat': 3093.0,
 'skoda': 4840.525423728814,
 'smart': 3629.1428571428573,
 'toyota': 4054.2,
 'volkswagen': 3403.743330266789,
 'volvo': 2716.035714285714}

In [142]:
unknown_damaged_cars_series = pd.Series(avg_damaged_by_brand_na)
print(unknown_damaged_cars_series)

audi             5819.794749
bmw              5993.334016
citroen          3320.525424
fiat             1856.700730
ford             2094.037464
hyundai          4652.794118
mazda            2671.704225
mercedes_benz    5751.973105
mini             6833.125000
mitsubishi       1796.897436
nissan           3159.750000
opel             1984.115917
peugeot          2214.206107
renault          1539.233871
seat             3093.000000
skoda            4840.525424
smart            3629.142857
toyota           4054.200000
volkswagen       3403.743330
volvo            2716.035714
dtype: float64


In [143]:
df2["unknown_damage"] = unknown_damaged_cars_series
df2

damaged_cars  undamaged_cars  unknown_damage
audi            3398.910345    11228.818658     5819.794749
bmw             3823.460641     9695.068355     5993.334016
citroen         1887.303571     4424.629454     3320.525424
fiat            1259.580000     3548.212251     1856.700730
ford            1574.253968     4856.317557     2094.037464
hyundai         2836.860465     6297.921875     4652.794118
mazda           1414.014493     5394.134831     2671.704225
mercedes_benz   4378.076923    10008.875866     5751.973105
mini            4749.090909    11141.760684     6833.125000
mitsubishi      1485.953488     4197.515021     1796.897436
nissan          2066.987500     6027.091518     3159.750000
opel            1535.222727     3787.832114     1984.115917
peugeot         1476.873134     3847.385945     2214.206107
renault         1261.066038     3216.078186     1539.233871
seat            1930.714286     5399.369416     3093.000000
skoda           3839.573770     7266.262609     4840.525424
smart           1558.769231     4068.869347     3629.142857
toyota          3721.933333     5628.972912     4054.200000
volkswagen      2376.165123     6743.192044     3403.743330
volvo           1893.150000     6020.924092     2716.035714

We can see from the above that cars which are undamaged command a significant premium over damaged cars. What is interesting is that listings which do not specify if the car is damaged or not have prices that lie in between the two values. That would suggest that if your car is undamaged, you could be losing money by leaving this attribute undefined. 

Another interesting observation is the price difference between damaged cars for some manufacturers. Take Audi for example. The price difference between a damaged and undamaged car on average is almost $8,000. Whereas Toyota has the smallest gap between damaged and undamaged cars. This may suggest that Audis are quite expensive to fix or have common major issues with them that are costly to repair. Toyotas on the other hand may be more reliable and could potentially be cheaper to repair.